File: sdk/dina-js/src/contract.ts
import type { Address, Hash, SpendingStats } from './types';
import { DinaClient } from './client';
import { DinaWallet } from './wallet';

/**
 * Generic smart contract interaction wrapper.
 *
 * Provides `call` (state-changing) and `view` (read-only) methods
 * that map to the Dina WASM contract ABI.
 */
export class DinaContract {
  constructor(
    public readonly address: Address,
    protected readonly client: DinaClient
  ) {}

  /**
   * Execute a state-changing contract method.
   * Requires a wallet for signing. Returns the transaction hash.
   */
  async call(
    method: string,
    args: unknown,
    wallet?: DinaWallet,
    usdcAttached?: bigint
  ): Promise<Hash> {
    if (!wallet) {
      throw new Error('Wallet required for state-changing calls');
    }
    return this.client.callContract(wallet, {
      contract: this.address,
      method,
      args,
      usdcAttached,
    });
  }

  /**
   * Execute a read-only view method. No wallet or gas needed.
   */
  async view(method: string, args: unknown): Promise<unknown> {
    return this.client['rpc']('dina_viewContract', [
      this.address,
      method,
      args,
    ]);
  }

  /** Create a DRC-1 token contract helper. */
  static token(address: Address, client: DinaClient): TokenContract {
    return new TokenContract(address, client);
  }

  /** Create a DRC-101 agent wallet contract helper. */
  static agentWallet(
    address: Address,
    client: DinaClient
  ): AgentWalletContract {
    return new AgentWalletContract(address, client);
  }
}

/**
 * DRC-1 Token standard contract — fungible token interface.
 * Similar to ERC-20 but for Dina Network.
 */
export class TokenContract extends DinaContract {
  /** Get token balance for an address. */
  async balanceOf(owner: Address): Promise<bigint> {
    const result = await this.view('balance_of', { account: owner });
    return BigInt(result as string);
  }

  /** Transfer tokens to another address. */
  async transfer(
    wallet: DinaWallet,
    to: Address,
    amount: bigint
  ): Promise<Hash> {
    return this.call(
      'transfer',
      { to, amount: amount.toString() },
      wallet
    );
  }

  /** Approve a spender to transfer tokens on your behalf. */
  async approve(
    wallet: DinaWallet,
    spender: Address,
    amount: bigint
  ): Promise<Hash> {
    return this.call(
      'approve',
      { spender, amount: amount.toString() },
      wallet
    );
  }

  /** Check how much a spender is allowed to spend on behalf of an owner. */
  async allowance(owner: Address, spender: Address): Promise<bigint> {
    const result = await this.view('allowance', { owner, spender });
    return BigInt(result as string);
  }

  /** Get the total token supply. */
  async totalSupply(): Promise<bigint> {
    const result = await this.view('total_supply', {});
    return BigInt(result as string);
  }
}

/**
 * DRC-101 Agent Wallet contract — AI agent spending wallet.
 * Provides spending limits, emergency controls, and stats.
 */
export class AgentWalletContract extends DinaContract {
  /** Execute a transfer from the agent wallet. */
  async executeTransfer(
    wallet: DinaWallet,
    to: Address,
    amount: bigint
  ): Promise<Hash> {
    return this.call(
      'execute_transfer',
      { to, amount: amount.toString() },
      wallet
    );
  }

  /** Get spending statistics for the agent wallet. */
  async spendingStats(): Promise<SpendingStats> {
    const raw = (await this.view('spending_stats', {})) as {
      totalSpent: string;
      transactionCount: number;
      lastTransaction: number;
      dailyLimit: string;
      dailySpent: string;
    };
    return {
      totalSpent: BigInt(raw.totalSpent),
      transactionCount: raw.transactionCount,
      lastTransaction: raw.lastTransaction,
      dailyLimit: BigInt(raw.dailyLimit),
      dailySpent: BigInt(raw.dailySpent),
    };
  }

  /** Halt all agent wallet activity immediately. */
  async emergencyStop(wallet: DinaWallet): Promise<Hash> {
    return this.call('emergency_stop', {}, wallet);
  }

  /** Resume agent wallet activity after an emergency stop. */
  async resume(wallet: DinaWallet): Promise<Hash> {
    return this.call('resume', {}, wallet);
  }
}

File: sdk/dina-py/src/dina/contract.py
"""Smart contract interaction helpers for the Dina Network."""

from __future__ import annotations

from typing import Any, Optional

from .client import DinaClient, DinaError
from .types import Address, Hash, TransactionReceipt
from .wallet import DinaWallet


class DinaContract:
    """Wrapper for interacting with a deployed Dina smart contract.

    Provides convenience methods for calling state-mutating and
    read-only (view) methods on a contract.

    Args:
        address: The on-chain address of the deployed contract.
        client: A connected DinaClient instance.
    """

    def __init__(self, address: Address, client: DinaClient) -> None:
        self.address = address
        self._client = client

    def call(
        self,
        method: str,
        args: Optional[dict[str, Any]] = None,
        wallet: Optional[DinaWallet] = None,
        usdc: int = 0,
        wait: bool = False,
        timeout: float = 30.0,
    ) -> Hash | TransactionReceipt:
        """Call a state-mutating method on the contract.

        Args:
            method: The contract method name.
            args: Keyword arguments to pass to the method.
            wallet: The wallet to sign the transaction. Required for
                state-mutating calls.
            usdc: Micro-USDC to send with the call (default 0).
            wait: If True, wait for the transaction to be confirmed and
                return the receipt instead of the hash.
            timeout: Seconds to wait if ``wait=True`` (default 30).

        Returns:
            The transaction hash (str) or a TransactionReceipt if
            ``wait=True``.

        Raises:
            ValueError: If no wallet is provided.
        """
        if wallet is None:
            raise ValueError("A wallet is required for state-mutating calls")

        tx_hash = self._client.call_contract(
            wallet=wallet,
            contract=self.address,
            method=method,
            args=args or {},
            usdc=usdc,
        )

        if wait:
            return self._client.wait_for_transaction(tx_hash, timeout=timeout)
        return tx_hash

    def view(self, method: str, args: Optional[dict[str, Any]] = None) -> Any:
        """Call a read-only (view) method on the contract.

        View calls do not require a wallet or transaction -- they query
        the current contract state via the RPC node.

        Args:
            method: The contract method name.
            args: Keyword arguments to pass to the method.

        Returns:
            The deserialized return value from the contract.
        """
        result = self._client._rpc(
            "contract.view",
            {
                "contract": self.address,
                "method": method,
                "args": args or {},
            },
        )
        return result

    def __repr__(self) -> str:
        return f"DinaContract(address={self.address!r})"


class TokenContract(DinaContract):
    """Helper for DRC-1 token contracts (fungible tokens like USDC on Dina).

    Provides typed wrappers around standard DRC-1 methods:
    name, symbol, decimals, total_supply, balance_of, transfer, approve,
    allowance.
    """

    def name(self) -> str:
        """Get the token name."""
        return str(self.view("name"))

    def symbol(self) -> str:
        """Get the token symbol."""
        return str(self.view("symbol"))

    def decimals(self) -> int:
        """Get the number of decimal places."""
        return int(self.view("decimals"))

    def total_supply(self) -> int:
        """Get the total supply in smallest units."""
        return int(self.view("total_supply"))

    def balance_of(self, owner: Address) -> int:
        """Get the token balance for an address.

        Args:
            owner: The address to query.

        Returns:
            Token balance in smallest units.
        """
        return int(self.view("balance_of", {"owner": owner}))

    def transfer(
        self,
        wallet: DinaWallet,
        to: Address,
        amount: int,
        wait: bool = False,
    ) -> Hash | TransactionReceipt:
        """Transfer tokens to another address.

        Args:
            wallet: Sender's wallet.
            to: Recipient address.
            amount: Amount in smallest units.
            wait: Wait for confirmation.

        Returns:
            Transaction hash or receipt.
        """
        return self.call("transfer", {"to": to, "amount": amount}, wallet=wallet, wait=wait)

    def approve(
        self,
        wallet: DinaWallet,
        spender: Address,
        amount: int,
        wait: bool = False,
    ) -> Hash | TransactionReceipt:
        """Approve a spender to transfer tokens on behalf of the caller.

        Args:
            wallet: Token owner's wallet.
            spender: Address to approve.
            amount: Maximum amount the spender can transfer.
            wait: Wait for confirmation.

        Returns:
            Transaction hash or receipt.
        """
        return self.call(
            "approve", {"spender": spender, "amount": amount}, wallet=wallet, wait=wait
        )

    def allowance(self, owner: Address, spender: Address) -> int:
        """Get the remaining allowance for a spender.

        Args:
            owner: Token owner address.
            spender: Approved spender address.

        Returns:
            Remaining allowance in smallest units.
        """
        return int(self.view("allowance", {"owner": owner, "spender": spender}))


class AgentWalletContract(DinaContract):
    """Helper for DRC-101 agent wallet contracts.

    DRC-101 contracts hold funds on behalf of an AI agent and expose
    methods for deposits, withdrawals, task execution, and
    operator management.
    """

    def owner(self) -> Address:
        """Get the wallet owner address."""
        return str(self.view("owner"))

    def agent_id(self) -> str:
        """Get the agent identifier."""
        return str(self.view("agent_id"))

    def balance(self) -> int:
        """Get the USDC balance held by the agent wallet."""
        return int(self.view("balance"))

    def is_operator(self, address: Address) -> bool:
        """Check if an address is an authorized operator.

        Args:
            address: The address to check.

        Returns:
            True if the address is an operator.
        """
        return bool(self.view("is_operator", {"address": address}))

    def deposit(
        self,
        wallet: DinaWallet,
        amount: int,
        wait: bool = False,
    ) -> Hash | TransactionReceipt:
        """Deposit USDC into the agent wallet.

        Args:
            wallet: Depositor's wallet.
            amount: Micro-USDC to deposit.
            wait: Wait for confirmation.

        Returns:
            Transaction hash or receipt.
        """
        return self.call("deposit", {}, wallet=wallet, usdc=amount, wait=wait)

    def withdraw(
        self,
        wallet: DinaWallet,
        amount: int,
        to: Optional[Address] = None,
        wait: bool = False,
    ) -> Hash | TransactionReceipt:
        """Withdraw USDC from the agent wallet.

        Args:
            wallet: Must be the owner or an operator.
            amount: Micro-USDC to withdraw.
            to: Destination address (defaults to wallet address).
            wait: Wait for confirmation.

        Returns:
            Transaction hash or receipt.
        """
        args: dict[str, Any] = {"amount": amount}
        if to is not None:
            args["to"] = to
        return self.call("withdraw", args, wallet=wallet, wait=wait)

    def execute_task(
        self,
        wallet: DinaWallet,
        task_id: str,
        payload: dict[str, Any],
        max_spend: int = 0,
        wait: bool = False,
    ) -> Hash | TransactionReceipt:
        """Execute an agent task, optionally spending USDC.

        Args:
            wallet: Must be the owner or an operator.
            task_id: Unique identifier for the task.
            payload: Task execution parameters.
            max_spend: Maximum micro-USDC the task can spend.
            wait: Wait for confirmation.

        Returns:
            Transaction hash or receipt.
        """
        args = {
            "task_id": task_id,
            "payload": payload,
            "max_spend": max_spend,
        }
        return self.call("execute_task", args, wallet=wallet, wait=wait)

    def add_operator(
        self,
        wallet: DinaWallet,
        operator: Address,
        wait: bool = False,
    ) -> Hash | TransactionReceipt:
        """Add an authorized operator to the agent wallet.

        Args:
            wallet: Must be the owner.
            operator: Address to authorize.
            wait: Wait for confirmation.

        Returns:
            Transaction hash or receipt.
        """
        return self.call("add_operator", {"operator": operator}, wallet=wallet, wait=wait)

    def remove_operator(
        self,
        wallet: DinaWallet,
        operator: Address,
        wait: bool = False,
    ) -> Hash | TransactionReceipt:
        """Remove an operator from the agent wallet.

        Args:
            wallet: Must be the owner.
            operator: Address to remove.
            wait: Wait for confirmation.

        Returns:
            Transaction hash or receipt.
        """
        return self.call("remove_operator", {"operator": operator}, wallet=wallet, wait=wait)
